Snakemake (#1978)
CSCfi/csc-user-guide: docs/support/tutorials/snakemake-puhti.md

## Running Snakemake workflow on Puhti

Snakemake workflow is one of the popular scientific workflows in the bioinformatics community although the workflow manager itself can be used in other scientific disciplines as well. The Snakemake enables scalable and reproducible scientific pipelines by chaining a series of rules in a fully-specified software environment. Snakemake software is available as a module in Puhti supercomputing environment. Also, a CSC user can easily install it in his/her own disk space (e.g., in /projappl directory) if a specific version of Snakemake is desired. The following toy example illustrates how a Snakemake workflow can be deployed at CSC.

> Please make sure that you have a [user account at CSC](../../accounts/how-to-create-new-user-account.md) and are a member of a project which [has access to the Puhti service](../../accounts/how-to-add-service-access-for-project.md) before start running workflows on Puhti. One should avoid launching heavy Snakemake workflow jobs on login nodes and use interactive or batch jobs instead. More information on using interactive jobs can be found on 

Reviewer: ```suggestion
Please make sure that you have a [user account at CSC](../../accounts/how-to-create-new-user-account.md) and are a member of a project which [has access to the Puhti service](../../accounts/how-to-add-service-access-for-project.md) before start running workflows on Puhti. One should avoid launching heavy Snakemake workflow jobs on **login nodes** and use interactive or batch jobs instead. More information on using interactive jobs can be found on 
```

```bash

rule all:
        input: "CAPITAL_CASE.txt"

rule say_hello:
        output: "smaller_case.txt"
        shell:
                """
                echo "hello-world greetings from csc to snakemake community !" > smaller_case.txt
                """
rule capitalise:
        input: "smaller_case.txt"
        output: "CAPITAL_CASE.txt"
        shell:
                """
                tr '[:lower:]' '[:upper:]' < {input} > {output}
                """
```
As snakemake is installed as a module on Puhti, one can simply load it without needing to install it separately. Here is a bash script for running the above toy example:
```bash
module load snakemake/7.17.1
snakemake --help   #  to get information on more options.

# The following command shows how to run a snakemake workflow on a cluster using slurm executor
snakemake -s Snakefile \ # the Snakefile is the default file name; no need specify with -s flag
 -j 1  \     # this will execute up to 3 tasks in parallel)       
--latency-wait 60 \  # snakemake to wait up to 60 seconds after a job completes for the output files to become available.

> --cluster "sbatch -t 10  --account=project_xxx --job-name=hello-world --tasks-per-node=1 --cpus-per-task=1 

Reviewer: ```suggestion
--cluster "sbatch --time 10  --account=project_xxx --job-name=hello-world --tasks-per-node=1 --cpus-per-task=1 
```

> --mem-per-cpu=4000 -p test"

Reviewer: ```suggestion
--mem-per-cpu=4000 --partition test"
```

# cluster option to execute the workflow on slurm
```
Finally, you can run the workflow in the bash script (e.g., run_snakemak.sh) by submitting in the job in the interactive node:
```bash

> sinteractive -c 2 -m 10000 # type this command on login node

Reviewer: ```suggestion
sinteractive --cores 2 --mem 10000 # type this command on login node to start an interactive session with 2 CPU cores and 10 Gb of memory
```

bash run_snakemake.sh   # run the workflow
```

> !!! note
>     If you are using cluster configuration for snakemake workflow,  
>     please note that cluster configurations are no longer supported on recent versions of snakemake. Please consider using   [profiles](https://snakemake.readthedocs.io/en/stable/executing/cli.html#profiles) instead.

Reviewer: ```suggestion
!!! note
    Please note that **cluster** configurations are no longer supported on recent versions of snakemake. Please consider using   [profiles](https://snakemake.readthedocs.io/en/stable/executing/cli.html#profiles) instead.
```

!!! note
    Scaling up of your jobs using slurm should be done carefully to
    avoid unnecessary overload on slurm accounting database due to a large number of small jobs. Consider either configuring **localrules** cautiously with slurm or using Hyperqueue executor. 

### Running snakemake with python packages installed *via.* tykky wrapper

> Conda installations should not be performed directly on Puhti. [Tykky container wrapper tool](../../computing/containers/tykky.md) instead be used to install python packages in setting up your compute environment. The wrapper tool installs applications inside of a singularity container and thus  facilitates better performance in terms of faster startup times, reduced IO load, and reduced number of files on parallel filesystems. 

Reviewer: Here could be a sentence first telling about how snakemake supports custom environments created for each rule (and link eg to this page: https://snakemake.readthedocs.io/en/stable/tutorial/additional_features.html#automatic-deployment-of-software-dependencies) leading to that this use is not possible on Puhti due to "no direct conda installations allowed" -policy.

Author: I added a sentence to use one tykky env for whole workflow

> Conda installations should not be performed directly on Puhti. [Tykky container wrapper tool](../../computing/containers/tykky.md) instead be used to install python packages in setting up your compute environment. The wrapper tool installs applications inside of a singularity container and thus  facilitates better performance in terms of faster startup times, reduced IO load, and reduced number of files on parallel filesystems. 

Reviewer: ```suggestion
Conda installations should not be performed directly on Puhti.  CSC instead provides the [Tykky container wrapper tool](../../computing/containers/tykky.md) which can instead be used to install python packages to set up your own compute environment. The wrapper tool installs applications inside of a singularity container and thus  facilitates better performance in terms of faster startup times, reduced IO load, and reduced number of files on parallel filesystems. 
```

Here is an example of tykky-based custom installation for conda packages (**note**: make sure to edit with the correct CSC project name and user name as needed):

```bash
# start an interactive session once you are in login node

> sinteractive -c 8 -m 30000 -d 100  # this command requests a compute node with 8 cores, 30 GB memory and 100 GB local disk space; change settings as needed

Reviewer: ```suggestion
sinteractive --cores 8 --mem 30000 --tmp 100  # this command requests a compute node with 8 cores, 30 GB memory and 100 GB local disk space; change settings as needed
```

[CSC documentation](../../computing/running/interactive-usage.md).

Let's use the following toy Snakemake script, Snakefile (with a capital S and no file extension), for the illustration:

```bash

rule all:
        input: "CAPITAL_CASE.txt"

rule say_hello:
        output: "smaller_case.txt"
        shell:
                """
                echo "hello-world greetings from csc to snakemake community !" > smaller_case.txt
                """
rule capitalise:
        input: "smaller_case.txt"
        output: "CAPITAL_CASE.txt"
        shell:
                """
                tr '[:lower:]' '[:upper:]' < {input} > {output}
                """
```
As snakemake is installed as a module on Puhti, one can simply load it without needing to install it separately. Here is a bash script for running the above toy example:
```bash

> module load snakemake/7.17.1

Reviewer 2: General idea: should we consider upgrading to 8.X.X? We could then just used the newest version of docs when it comes to slurm/hyperqueue integration and avoid any misunderstandings.

Author: This is something I considered earlier. Plan is to install recent version of snakemake (as  non-containerised ) module and make it compatible with other container environments that users may come up with.  Yes. this needs bit of testing.

Reviewer 2: I am leaving it up to your decision :) Feel free to close this thread.

Author: No worries. I pushed one more version with recent version of snakemake.

snakemake --help   #  to get information on more options.

# The following command shows how to run a snakemake workflow on a cluster using slurm executor
snakemake -s Snakefile \ # the Snakefile is the default file name; no need specify with -s flag

> --jobs 1  \     # this will execute up to 3 tasks in parallel)       

Reviewer 2: Why "up to 3 tasks in parallel"? Not only one task in parallel?

+ typo at the end

--latency-wait 60 \  # snakemake to wait up to 60 seconds after a job completes for the output files to become available.
--cluster "sbatch --time 10  --account=project_xxx --job-name=hello-world --tasks-per-node=1 --cpus-per-task=1 
--mem-per-cpu=4000 --partition test"
# cluster option to execute the workflow on slurm
```
Finally, you can run the workflow in the bash script (e.g., run_snakemak.sh) by submitting in the job in the interactive node:
```bash
sinteractive --cores 2 --mem 10000 # type this command on login node to start an interactive session with 2 CPU cores and 10 Gb of memory, after providing your project when prompted, you can run the bash script like this:
bash run_snakemake.sh   # run the workflow
```
!!! note

>     Please note that **cluster** configurations are no longer supported on recent versions of snakemake. Please consider using   [profiles](https://snakemake.readthedocs.io/en/stable/executing/cli.html#profiles) instead.

Reviewer 2: Maybe we can link this as well: https://snakemake.github.io/snakemake-plugin-catalog/plugins/executor/slurm.html. This page contains more information about current snakemake cluster configuration. If not, it would be nice to be more specific with versioning: this tutorial works for snakemake 7.X.X but not for snakemake 8.X.X.

Author: In this current version, we wrote this issue as a note. although we don't recommend slurm, nice to test especially hyperqueue executor syntax with 8.X.X. version of snakemake

Author: when workflows run at scale, especially with lot of small jobs, would cause some extra overload on slurm cluster (actually on slurm accounting DB). That is why, we want to avoid slurm and use HQ if possible.  One can of course carefully write local rules in snakemake script to minimise these small jobs.
